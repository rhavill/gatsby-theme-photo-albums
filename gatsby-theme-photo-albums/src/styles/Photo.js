/** @jsx jsx */
import { jsx } from 'theme-ui'
import PropTypes from 'prop-types'

const Photo = ({children, height}) => 
  <div sx={{variant: 'photoPage', height}}>
    {children}
  </div>

Photo.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  height: PropTypes.number.isRequired,
}

export default Photo