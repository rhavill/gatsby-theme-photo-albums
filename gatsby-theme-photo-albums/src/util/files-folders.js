const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')
const {prependbaseUrl} = require('./url-text')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(currentPath, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

/**
 * Accepts an array of gatsby-source-filesystem objects that must have a
 * relativePath property. Adds a url property to all of the objects.
 */
const addUrlProps = curry((baseUrl, filesOrDirectories) => 
  filesOrDirectories.map(fileOrDirectory => ({
    ...fileOrDirectory,
    url: prependbaseUrl(baseUrl, fileOrDirectory.relativePath)
  })))

module.exports = {
  addUrlProps,
  getChildPaths,
  getPagerData,
}