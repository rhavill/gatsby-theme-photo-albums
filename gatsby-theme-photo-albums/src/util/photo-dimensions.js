export default (windowDimensions, pageContext) => {
  const {width, height} = windowDimensions
  const ratio = pageContext.width / pageContext.height
  if (pageContext.width < width && pageContext.height < height)
    return {maxWidth: width, maxHeight: height}
  const excessWidthFactor = (pageContext.width - width) / width
  const excessHeightFactor = (pageContext.height - height) / height
  return excessWidthFactor > excessHeightFactor
    ? {maxWidth: width, maxHeight: width / ratio}
    : {maxWidth: height * ratio, maxHeight: height}
}