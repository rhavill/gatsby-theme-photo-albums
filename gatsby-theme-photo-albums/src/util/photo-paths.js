const forEach = require('ramda/src/forEach')
const keys = require('ramda/src/keys')
const match = require('ramda/src/match')
const {indexedForEach, groupByProp} = require('./ramda-utils')
const {removeFileExtension} = require('./url-text')
const groupByRelativeDirectory = groupByProp('relativeDirectory')

const getPhotoPathsWithPages = (photosPerPage, fileData) => {
  const photoPaths = {}
  const groupedFiles = groupByRelativeDirectory(fileData)
  forEach(relativeDirectory => indexedForEach(
    (file, i) => {
      const pageNumber = Math.ceil((i + 1) / photosPerPage)
      photoPaths[file.url] = getPhotoPathWithPage(
        pageNumber, file.url
      )
    }, 
    groupedFiles[relativeDirectory]
  ), keys(groupedFiles))
  return photoPaths
}

const getPhotoPathWithPage = (pageNumber, url) => {
  url = removeFileExtension(url)
  const matches = match(/^(.*\/)([^/]+)$/, url)
  if (matches && matches[1] && matches[2]) {
    const path =  `${matches[1]}` + (pageNumber > 1 ? `${pageNumber}/` : '') + matches[2]
    return path
  }
  return url
}

module.exports = {getPhotoPathsWithPages, getPhotoPathWithPage}