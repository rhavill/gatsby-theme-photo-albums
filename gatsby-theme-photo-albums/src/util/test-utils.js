import mergeDeepRight from 'ramda/src/mergeDeepRight'
import React from 'react'
import PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'emotion-theming'

import theme from '../theme'

const create = (element, options) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{element}</ThemeProvider>, options
  )
}

const customRenderer = mergeDeepRight(renderer, {create})

const Wrapper = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
}
Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options })

// re-export everything from React Testing Library
export * from '@testing-library/react'

// override react testing library render method and custom react-test-renderer r
export { customRender as render, customRenderer as renderer }