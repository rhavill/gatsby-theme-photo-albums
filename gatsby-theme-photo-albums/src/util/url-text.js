const compose = require('ramda/src/compose')
const concat = require('ramda/src/concat')
const curry = require('ramda/src/curry')
const replace = require('ramda/src/replace')

const toTitleCase = text => 
  text.toLowerCase()
    .split(/[-_]+/)
    .map((s) => s.charAt(0).toUpperCase() + s.substring(1))
    .join(' ')

const removeFileExtension = replace(/^(.+)\.[^.]+$/, '$1')

const pathToFile = replace(/.*\/([^/]+$)/, '$1')

const pathToFileTitle = compose(toTitleCase, removeFileExtension, pathToFile)

const prependBaseUrl = curry((baseUrl, relativePath) => {
  // Make sure baseUrl starts and ends with slash (for consistency)
  const base = ensureLeadingAndTrailingSlash(baseUrl)
  return concat(base, relativePath)
})

const ensureTrailingSlash = text => text.slice(-1) === '/' ? text : text + '/'

const ensureLeadingSlash = text => text.charAt(0) === '/' ? text : '/' + text

const ensureLeadingAndTrailingSlash = compose(ensureLeadingSlash, ensureTrailingSlash)

module.exports = {
  pathToFileTitle,
  prependBaseUrl,
  removeFileExtension,
  toTitleCase,
  ensureLeadingAndTrailingSlash,
}