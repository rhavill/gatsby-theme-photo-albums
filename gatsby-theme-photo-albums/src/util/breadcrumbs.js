import compose from 'ramda/src/compose'
import {removeFileExtension, toTitleCase} from './url-text'

const breadcrumbs = (path) => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' 
        ? '' 
        : compose(toTitleCase, removeFileExtension)(folder),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs