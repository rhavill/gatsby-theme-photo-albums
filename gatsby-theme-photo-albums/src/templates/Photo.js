import compose from 'ramda/src/compose'
import React, {useRef} from 'react'
import PropTypes from 'prop-types'
import {graphql} from 'gatsby'
import Img from 'gatsby-image'
import Layout from '../components/layout/Layout'
import {pathToFileTitle, removePathPrefix, removeFileExtension} from '../util/url-text'
import useWindowDimensions from '../hooks/use-window-dimensions'
import photoDimensions from '../util/photo-dimensions'

const Photo =  ({data, path, pageContext}) => {
  
  const pathPrefix = data.site.pathPrefix
  path = compose(
    removeFileExtension, removePathPrefix(pathPrefix), decodeURI
  )(path)
  const title = pathToFileTitle(path)
  const ref = useRef(null)
  const windowDimensions = useWindowDimensions()
  const dimensions = photoDimensions(windowDimensions, pageContext)
console.log('window', windowDimensions, 'dim', dimensions)
  return (
    <Layout path={path}>
      <div className='photo-page' data-testid={path} ref={ref}>
        <Img fixed={data.photo.childImageSharp.fixed} alt={title} title={title}
          loading='eager' css={dimensions}/>
      </div>
    </Layout>
  )
}

export const query = graphql`
  query photoQuery($relativePath: String!, $width: Int!, $height: Int!) {
    site {
      pathPrefix
    }
    photo: file(relativePath: {eq: $relativePath}) {
      childImageSharp {
        fixed(width: $width, height: $height) {
          ...GatsbyImageSharpFixed
        }
      }
    }
  }
`

Photo.propTypes = {
  path: PropTypes.string.isRequired,
  data: PropTypes.shape({
    site: PropTypes.shape({
      pathPrefix: PropTypes.string.isRequired,
    }),
    photo: PropTypes.shape({
      childImageSharp: PropTypes.shape({
        fixed: PropTypes.shape({
          src: PropTypes.string.isRequired,
          srcSet: PropTypes.string.isRequired,
          width: PropTypes.number.isRequired,
          height: PropTypes.number.isRequired,
        }).isRequired
      })
    })
  }),
  pageContext: PropTypes.shape({
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
  }).isRequired
}

export default Photo