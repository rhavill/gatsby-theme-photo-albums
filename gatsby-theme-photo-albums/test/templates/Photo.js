import React from 'react'
import {cleanup, render, renderer} from '../../src/util/test-utils'
import Photo from '../../src/templates/Photo'
import photoData from '../../test-data/source-filesystem-photo-data'
import {pathToFileTitle, removeFileExtension} from '../../src/util/url-text'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('Photo', () => {
  const pageContext = {
    width: 1024,
    height: 768,
    file: {
      previousUrl: '/prev',
      nextUrl: '/next',
      parentUrl: '/parent',
    },
  }
  // This test is being disabled because react-zoom-pan-pinch is causing an error:
  // "TypeError: Cannot read property 'style' of null"
  // it('renders correctly', () => {
  //   const tree = renderer
  //     .create(<Photo path='/base/1' data={photoData.data} pageContext={pageContext} />)
  //     .toJSON()
  //   expect(tree).toMatchSnapshot()
  // })
  it('displays image with "title" attribute containing a title-case version of the photo file name', () => {
    const path = removeFileExtension(photoData.data.photo.childImageSharp.fixed.src)
    const {getAllByTitle} = render(
      <Photo path={path} data={photoData.data} pageContext={pageContext} />
    )
    const expected = pathToFileTitle(path)
    expect(getAllByTitle(expected).length).toBe(1)
  })
  it('passes decoded url to child components', () => {
    const path = '/san-sebasti%c3%a1n/san-sebasti%c3%a1n'  
    const {getAllByTestId} = render(
      <Photo path={path} data={photoData.data} pageContext={pageContext} />
    )
    expect(getAllByTestId('/san-sebastián/san-sebastián').length).toBe(1)
  })
  it('passes path to child components without pathPrefix config variable', () => {
    const path = '/path-prefix/top-level/file.jpg'  
    const {getAllByTestId} = render(
      <Photo path={path} data={photoData.data} pageContext={pageContext} />
    )
    expect(getAllByTestId('/top-level/file').length).toBe(1)
  })
})