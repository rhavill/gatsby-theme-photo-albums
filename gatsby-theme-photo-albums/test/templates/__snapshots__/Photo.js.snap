// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo renders correctly 1`] = `
<div
  className="css-1qldoov-Layout"
>
  <div
    className="css-1jhlbdc-Header"
  >
    <nav
      className="css-1be44vx-Menu"
    >
      <label
        htmlFor="menu-checkbox"
      >
        ☰
      </label>
      <input
        defaultChecked={false}
        id="menu-checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      <ul
        css={
          Object {
            "display": "none",
          }
        }
      >
        <a
          href="/"
        >
          <li>
            <span
              className="first-breadcrumb"
            >
              Home
            </span>
          </li>
        </a>
        <a
          href="/base"
        >
          <li>
            Base
          </li>
        </a>
        <a
          href="/base/1"
        >
          <li>
            1
          </li>
        </a>
      </ul>
    </nav>
  </div>
  <div
    className="css-15phz5z-MainContent"
  >
    <div
      data-testid="/base/1"
      onMouseDown={[Function]}
    >
      <div
        className="css-3q5667-Photo"
      >
        <div
          className="photo-navigation previous"
        >
          <div>
            <a
              href="/prev"
            >
              ◄
            </a>
          </div>
        </div>
        <div
          className=" gatsby-image-wrapper"
          style={
            Object {
              "display": "inline-block",
              "height": 250,
              "overflow": "hidden",
              "position": "relative",
              "width": 250,
            }
          }
        >
          <picture>
            <source
              srcSet="/static/blah/blah/1.jpg 1x"
            />
            <img
              alt="1"
              draggable={false}
              height={250}
              loading="eager"
              onLoad={[Function]}
              src="/static/blah/blah/1.jpg"
              srcSet="/static/blah/blah/1.jpg 1x"
              style={
                Object {
                  "height": "100%",
                  "left": 0,
                  "objectFit": "cover",
                  "objectPosition": "center",
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                  "transition": "opacity 500ms",
                  "width": "100%",
                }
              }
              title="1"
              width={250}
            />
          </picture>
          <noscript
            dangerouslySetInnerHTML={
              Object {
                "__html": "<picture><source srcset=\\"/static/blah/blah/1.jpg 1x\\" /><img loading=\\"eager\\" width=\\"250\\" height=\\"250\\" srcset=\\"/static/blah/blah/1.jpg 1x\\" src=\\"/static/blah/blah/1.jpg\\" alt=\\"1\\" title=\\"1\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
              }
            }
          />
        </div>
        <div
          className="photo-navigation parent"
        >
          <div>
            <a
              href="/parent"
            >
              ⯅
            </a>
          </div>
        </div>
        <div
          className="photo-navigation next"
        >
          <div>
            <a
              href="/next"
            >
              ►
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
